### ISSUE TITLE
Implicit boxing of large futures causes excessive monomorphization 

### ISSUE THREAD

**Issue by osiewicz** [stjepang/async-task#66]
`async-task` contains an optimisation for handling large futures in the definition of [spawn_unchecked](https://github.com/smol-rs/async-task/blob/master/src/runnable.rs#L516). This leads to excessive IR size, as one branch instantiates `RawTask` with Fut and the other does so with `Pin<Box<Fut>>`. This probably gets eliminated within LLVM (as the branch itself is trivial), but it's still a bummer that this cannot be truly determined at compile time. I took several stabs recently at getting rid of the unnecessary instantiation, without luck. I do understand why we need the boxing, but it'd be nice to not spend time on generating code we're gonna throw away in LLVM anyways.
Getting rid of large-future-boxing reduces the LLVM IR size for my medium-size (~1.5M LLVM IR lines in debug) crate by 7%. This is also replicated in examples from this crate:
| Example name| LLVM IR before | LLVM IR after | % of original |
|---------------|----------------|---------------|--------------|
| spawn | 18276 | 15631 | 85.52% |
| spawn-local | 39801 | 34537 | 86.77% |
| spawn-on-thread | 18667 | 16031 | 85.87% |
| with-metadata | 32834 | 24887 | 75.79% |

Related: https://github.com/rust-lang/rust/issues/85836

**Comment by notgull** [comment]
Does this affect release builds too? I would be okay with doing something about this if it affects compile times. The crate itself compiles in 0.22s on my machine, but if the issue is monomorphisation I can see why it would be an issue.

Ideally any compile time improvements would be tested on a larger-scale crate that depends on `async-task`, like [`tide`](https://crates.io/crates/tide)'s examples. Before we commit to any new patterns we should benchmark against that.

One possible way would be to disable the `Box`ing branch on `debug_assertions` or with a specific feature flag. Again, I would like to see benchmarks before I commit to this.

**Comment by osiewicz** [comment]
Yeah, this does affect release builds as well. I wrote up a [trivial benchmark](https://gist.github.com/osiewicz/cd7358f1cc7cf50543cac1f4e0b85c53):
| | Debug| Release |
|--------|--------|--------|
| Baseline | 0.9s | 2s |
| Patched | 0.65s | 1.35s |
| % of Baseline | 72% | 67% | 
| LLVM Lines Baseline | 170310 | 144200 |
| LLVM Lines Patched | 85023 | 71760 |
| LLVM Lines % of Baseline | 50% | 50% |

You can increase the # of lines and the compile time will scale linearly, as on each line we essentially instantiate a new scheduler and future. If the user were to Box the scheduler, erasing the type, we would scale by the # of distinct return types instead.
I'm not sure if `tide` is a great benchmark there, as # of instantiations of `async-task` seems rather scarce in the examples, so changes are not likely to show up in timings. I think we could start with an isolated benchmark and move from there. 

I would err on the side of disabling the boxing branch with a specific feature flag (or rather, enabling the boxing with a feature flag). That feature could be enabled by default. 
Even if the user *always* boxes their futures, they have to pay the compile-time cost (as then the "boxing branch" tries to instantiate `Pin<Box<Pin<Box<F>>>>`). In that case being able to disable the implicit boxing altogether would be great for them.
I'm pretty positive that there's other avenues we could explore to improve the compile time further for downstream users. I'm glad that you're open to that. :)

**Comment by osiewicz** [comment]
Since we're now open source, just throwing it out there: the crate I was benchmarking against in this issue was [Zed](https://github.com/zed-industries/zed).

**Comment by james7132** [comment]
Could you test against the changes in https://github.com/smol-rs/async-task/pull/87 to see if it has a tangible impact on codegen? The changes there should minimize the number of duplicate instances of many of those functions, though it doesn't directly impact the codegen from boxing, it should share a lot of the generic functions between both instances.

**Comment by osiewicz** [comment]
@james7132 Thanks for taking this on (at least somewhat); here's the result for the benchmark I've shared in my last comment:
Dev master (a11c4b22cbcbbdbc4fa0ff62bdbffb430a9d3394):  160878 lines, 3319 copies
Release master (a11c4b22cbcbbdbc4fa0ff62bdbffb430a9d3394): 129978 lines, 2650 copies
Dev on your branch (b3abfa1811646eb06f82427ea74d9e77cb75a1ee): 131797 lines, 3095 copies
Release on your branch (b3abfa1811646eb06f82427ea74d9e77cb75a1ee): 123622 lines, 2439 copies

Not too shabby. Dev build time (clean) goes from 0.54s to 0.48, which is about 10%.
Release build time: 1.35s -> 1.12s: 20% or so?

**Comment by osiewicz** [comment]
Also, with your PR you're bumping MSRV to 1.75. If we were to bump to 1.79, we could use `const` block to effectively solve this issue, which could be a great gain.

**Comment by james7132** [comment]
> Also, with your PR you're bumping MSRV to 1.75. If we were to bump to 1.80, we could use `const` block to effectively solve this issue, which could be a great gain.

Can you clarify how const blocks would resolve this issue?

I was originally going to propose using an explicit in-place construction (via closures + `&mut MaybeUninit<T>`) solution to solve the original problem of avoiding stack overflows, which would get rid of the boxing branch entirely.

**Comment by osiewicz** [comment]
Sure. 
Starting with Rust 1.79, you could restructure this line:
https://github.com/smol-rs/async-task/blob/a11c4b22cbcbbdbc4fa0ff62bdbffb430a9d3394/src/runnable.rs#L516
into:
```
let ptr = if const { mem::size_of::<Fut>() >= 2048 } {
```
Done that way, the LLVM bitcode will be generated for just one of the branches (either `if` or `else`, depending on the size of a Future). This would effectively slash the size of LLVM bitcode in half for *any* use of async_task.

**Comment by james7132** [comment]
OK, I've made a few improvements to #87, and tried it with #88. I ran your benchmark and got these results in release mode:

|branch|LLVM Lines|Copies|% diff against master|
|:---|:--|:--|:--|
|master|102415|2649|0%|
|polymorphization|87630|2230|-14%|
|no-boxing|47276|1177|-53.8%
|merged|35983|826|-64.8%|
